File: Samples/C#/MaterialExport/MaterialExportPlugin.cs
/*  Material Export
 *
 *  This sample illustrates how to export 3D object material via the Renga API.
 *
 *  Copyright Renga Software LLC, 2018. All rights reserved.
 */

using System;
using System.Text;

namespace MaterialExport
{
  public class MaterialExportPlugin : Renga.IPlugin
  {
    private Renga.Application m_application;
    private Renga.ActionEventSource m_actionEventSource;

    public bool Initialize(string pluginFolder)
    {
      m_application = new Renga.Application();

      var ui = m_application.UI;

      var uiPanelExtension = ui.CreateUIPanelExtension();

      var action = ui.CreateAction();
      action.ToolTip = "Export Objects";

      m_actionEventSource = new Renga.ActionEventSource(action);
      m_actionEventSource.Triggered += OnExport3DObjectsMaterialsButtonClicked;

      uiPanelExtension.AddToolButton(action);

      ui.AddExtensionToPrimaryPanel(uiPanelExtension);

      return true;
    }

    public void Stop()
    {
      m_actionEventSource.Dispose();
    }

    private void OnExport3DObjectsMaterialsButtonClicked(object sender, EventArgs args)
    {
      string commandOutput = Export3DObjectsMaterials();

      var ui = m_application.UI;
      ui.ShowMessageBox(Renga.MessageIcon.MessageIcon_Info,
        "Export 3D objects materials result output", commandOutput);
    }

    private string Export3DObjectsMaterials()
    {
      var project = m_application.Project;

      var dataExporter = project.DataExporter;
      var object3dCollection = dataExporter.GetObjects3D();

      var model = project.Model;
      var modelObjects = model.GetObjects();

      var materialManager = project.MaterialManager;
      var layeredMaterialManager = project.LayeredMaterialManager;

      var result = new StringBuilder();
      result.AppendLine(object3dCollection.Count.ToString() + " objects exported:");
      result.AppendLine();

      for (int i = 0; i < object3dCollection.Count; i++)
      {
        var object3d = object3dCollection.Get(i);

        var modelObjectId = object3d.ModelObjectId;
        var modelObject = modelObjects.GetById(modelObjectId);

        result.AppendLine("Object: {" + (i + 1) + "}" + modelObject.Name);

        var objectWithMaterial = modelObject as Renga.IObjectWithMaterial;
        if (objectWithMaterial != null)
        {
          result.AppendLine("  Material type: One-layered");

          var material = materialManager.GetMaterial(objectWithMaterial.MaterialId);

          if (material != null)
            result.AppendLine("  Material name: " + material.Name);
          else
            result.AppendLine("  No material assigned");

          result.AppendLine();

          continue;
        }

        var objectWithLayeredMaterial = modelObject as Renga.IObjectWithLayeredMaterial;

        if (objectWithLayeredMaterial != null)
        {
          result.AppendLine("  Material type: Multi-layered");

          var layeredMaterial = layeredMaterialManager.GetLayeredMaterial(objectWithLayeredMaterial.LayeredMaterialId);

          if (layeredMaterial != null)
          {
            result.AppendLine("  Sandwich name: " + layeredMaterial.Name);

            var layers = layeredMaterial.Layers;

            result.AppendLine($"  Sandwich consists of: {layers.Count} layers");
            result.AppendLine();

            for (int l = 0; l < layers.Count; l++)
            {
              var layer = layers.Get(l);

              result.AppendLine("    Layer #" + l);
              result.AppendLine("    Name: " + layer.Material.Name);
              result.AppendLine("    Thickness: " + layer.Thickness);
            }
          }
          else
          {
            result.AppendLine("  No material assigned");
          }

          result.AppendLine();

          continue;
        }

        result.AppendLine("  Material type: None");
        result.AppendLine();
      }

      return result.ToString();
    }
  }
}

File: Samples/C++/MaterialExport/MaterialExportPlugin.cpp
/*  Sample 3: Material Export
 *
 *  This sample illustrates how to export 3D object material via the Renga API.
 *
 *  Copyright Renga Software LLC, 2018. All rights reserved.
 */

#include "MaterialExportPlugin.h"

#include <sstream>

#include <Renga/CreateApplication.hpp>


MaterialExportPlugin::MaterialExportPlugin()
{
  ::CoInitializeEx(nullptr, COINIT_APARTMENTTHREADED);
}

MaterialExportPlugin::~MaterialExportPlugin()
{
  ::CoUninitialize();
}

bool MaterialExportPlugin::initialize(const wchar_t* pluginDir)
{
  auto pApplication = Renga::CreateApplication();
  if (!pApplication)
    return false;

  if (auto pUI = pApplication->GetUI())
  {
    if (auto pUIPanelExtension = pUI->CreateUIPanelExtension())
    {
      if (auto pAction = pUI->CreateAction())
      {
        pAction->PutToolTip(L"Export objects materials");
        m_pExportMaterialsButtonClickHandler.reset(new Export3DObjectsMaterialsButtonClickHandler(pAction, pApplication));
        pUIPanelExtension->AddToolButton(pAction);
      }
      pUI->AddExtensionToPrimaryPanel(pUIPanelExtension);
    }
  }
  return true;
}

void MaterialExportPlugin::stop()
{
  m_pExportMaterialsButtonClickHandler.reset();
}


Export3DObjectsMaterialsButtonClickHandler::Export3DObjectsMaterialsButtonClickHandler(Renga::IActionPtr pAction, Renga::IApplicationPtr pApplication) :
  ActionEventHandler(pAction),
  m_pApplication(pApplication)
{
}


void Export3DObjectsMaterialsButtonClickHandler::OnTriggered()
{
  std::wstring command_output = export3DObjectsMaterials();

  MessageBox(NULL,
    (LPCWSTR)command_output.c_str(),
    (LPCWSTR)L"Export 3D objects materials result output",
    MB_ICONINFORMATION | MB_OK );
}

std::wstring Export3DObjectsMaterialsButtonClickHandler::export3DObjectsMaterials() const
{
  auto pProject = m_pApplication->GetProject();

  auto pDataExporter = pProject->GetDataExporter();
  auto pObject3DCollection = pDataExporter->GetObjects3D();

  auto pModel = pProject->GetModel();
  auto pModelObjectCollection = pModel->GetObjects();

  auto pMaterialManager = pProject->GetMaterialManager();
  auto pLayeredMaterialManager = pProject->GetLayeredMaterialManager();


  std::wstringstream result_output;
  result_output << pObject3DCollection->GetCount() << " objects exported:" << "\n\n";
  for (int i = 0; i < pObject3DCollection->GetCount(); ++i)
  {
    auto pObject3D = pObject3DCollection->Get(i);

    const auto modelObjectId = pObject3D->GetModelObjectId();
    auto pModelObject = pModelObjectCollection->GetById(modelObjectId);

    result_output << "Object: {" << i + 1 << "} " << static_cast<const wchar_t*>(pModelObject->GetName()) << "\n";

    Renga::IObjectWithMaterialPtr pObjectWithMaterial;
    pModelObject->QueryInterface(&pObjectWithMaterial);
    if (pObjectWithMaterial)
    {
      result_output << "  Material type: " << "One-layered" << "\n";
      if (auto pMaterial = pMaterialManager->GetMaterial(pObjectWithMaterial->GetMaterialId()))
        result_output << "  Material name: " << static_cast<const wchar_t*>(pMaterial->GetName()) << "\n";
      else
        result_output << "  No material assigned" << "\n";

      result_output << "\n";
      continue;
    }

    Renga::IObjectWithLayeredMaterialPtr pObjectWithLayeredMaterial;
    pModelObject->QueryInterface(&pObjectWithLayeredMaterial);
    if (pObjectWithLayeredMaterial)
    {
      result_output << "  Material type: " << "Multi-layered" << "\n";
      if (auto pLayeredMaterial = pLayeredMaterialManager->GetLayeredMaterialByIdGroupPair(pObjectWithLayeredMaterial->GetLayeredMaterialIdGroupPair()))
      {
        result_output << "  Sandwich name: " << static_cast<const wchar_t*>(pLayeredMaterial->GetName()) << "\n";

        auto pLayers = pLayeredMaterial->GetLayers();

        result_output << "  Sandwich consists of: " << pLayers->GetCount() << " layers" << "\n\n";
        for (int l = 0; l < pLayers->GetCount(); ++l)
        {
          auto pLayer = pLayers->Get(l);

          result_output << "    Layer #" << l << "\n";
          result_output << "    Name: " << static_cast<const wchar_t*>(pLayer->GetMaterial()->GetName()) << "\n";
          result_output << "    Thickness: " << pLayer->GetThickness() << "\n";
        }
      }
      else
      {
        result_output << "  No material assigned" << "\n";
      }
      result_output << "\n";
      continue;
    }

    result_output << "  Material type: " << "None" << "\n";
    result_output << "\n";
  }
  return result_output.str();
}
